applications/FluidDynamicsApplication/symbolic_generation/compressible_navier_stokes/compressible_navier_stokes_explicit_cpp_template_with_integration.cpp

```diff
@@ -0,0 +1,603 @@
+    const double &U_0_1 = data.U(0, 1);
+    const double &U_0_2 = data.U(0, 2);
+    const double &U_0_3 = data.U(0, 3);
+    const double &U_0_4 = data.U(0, 4);
+    const double &U_1_0 = data.U(1, 0);
+    const double &U_1_1 = data.U(1, 1);
+    const double &U_1_2 = data.U(1, 2);
+    const double &U_1_3 = data.U(1, 3);
+    const double &U_1_4 = data.U(1, 4);
+    const double &U_2_0 = data.U(2, 0);
+    const double &U_2_1 = data.U(2, 1);
+    const double &U_2_2 = data.U(2, 2);
+    const double &U_2_3 = data.U(2, 3);
+    const double &U_2_4 = data.U(2, 4);
+    const double &U_3_0 = data.U(3, 0);
+    const double &U_3_1 = data.U(3, 1);
+    const double &U_3_2 = data.U(3, 2);
+    const double &U_3_3 = data.U(3, 3);
+    const double &U_3_4 = data.U(3, 4);
+
+    // Hardcoded shape functions for linear tetrahedra element
+    // This is explicitly done to minimize the alocation and matrix acceses
+    // The notation N_i_j means shape function for node j in Gauss pt. i
+    const double N_0_0 = 0.58541020;
+    const double N_0_1 = 0.13819660;
+    const double N_0_2 = 0.13819660;
+    const double N_0_3 = 0.13819660;
+    const double N_1_0 = 0.13819660;
+    const double N_1_1 = 0.58541020;
+    const double N_1_2 = 0.13819660;
+    const double N_1_3 = 0.13819660;
+    const double N_2_0 = 0.13819660;
+    const double N_2_1 = 0.13819660;
+    const double N_2_2 = 0.58541020;
+    const double N_2_3 = 0.13819660;
+    const double N_3_0 = 0.13819660;
+    const double N_3_1 = 0.13819660;
+    const double N_3_2 = 0.13819660;
+    const double N_3_3 = 0.58541020;
+
+    // Hardcoded shape functions gradients for linear tetrahedra element
+    // This is explicitly done to minimize the matrix acceses
+    // The notation DN_i_j means shape function for node i in dimension j
+    const double &DN_DX_0_0 = data.DN_DX(0, 0);
+    const double &DN_DX_0_1 = data.DN_DX(0, 1);
+    const double &DN_DX_0_2 = data.DN_DX(0, 2);
+    const double &DN_DX_1_0 = data.DN_DX(1, 0);
+    const double &DN_DX_1_1 = data.DN_DX(1, 1);
+    const double &DN_DX_1_2 = data.DN_DX(1, 2);
+    const double &DN_DX_2_0 = data.DN_DX(2, 0);
+    const double &DN_DX_2_1 = data.DN_DX(2, 1);
+    const double &DN_DX_2_2 = data.DN_DX(2, 2);
+    const double &DN_DX_3_0 = data.DN_DX(3, 0);
+    const double &DN_DX_3_1 = data.DN_DX(3, 1);
+    const double &DN_DX_3_2 = data.DN_DX(3, 2);
+
+    //substitute_rhs_3D
+
+    // Here we assume that all the weights of the gauss points are the same so we multiply at the end by Volume/n_nodes
+    rRightHandSideVector *= data.volume / static_cast<double>(n_nodes);
+
+    KRATOS_CATCH("")
+}
+
+template <>
+void CompressibleNavierStokesExplicit<2>::AddExplicitContribution(const ProcessInfo &rCurrentProcessInfo)
+{
+    constexpr IndexType dim = 2;
+    constexpr IndexType n_nodes = 3;
+    constexpr IndexType block_size = 4;
+
+    // Calculate the explicit residual vector
+    VectorType rhs;

```

Comment:
I'd create an auxiliary method so the RHS can use bounded matrices.

Reply:
Done.